### PR TITLE
Alfoa/ga fucntions constants

### LINKED ISSUE #1665: [DEFECT] Genetic Algorithm doesn't recognize constants and functions
--------
Defect Description
--------

Genetic Algorithm optimizer doesn't recognize constants or functions.

##### What did you expect to see happen?
Expected the optimizer node to pass constants via multirun that could be accessed by an external model.

##### What did you see instead?
External model doesn't receive constants and throws an error when trying to access the constants.

##### Do you have a suggested fix for the development team?
Have GA send constants like gradient descent

**Describe how to Reproduce**
1. Put constant node into genetic algorithm optimization node/ MC sampler node in any of the GA tests
2. Have external model manipulate the constant
3. Get error AttributeError: 'Object' object has no attribute '[constant name]'
4. 

**Screenshots and Input Files**
Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

Added Constant:
image

Manipulated in external model:
image

Error:
image

**Platform (please complete the following information):**
 - OS: Windows
 - Version: [e.g. 22]
 - Dependencies Installation: CONDA
----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by aalfonsi**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1665
##### What are the significant changes in functionality due to this change request?

Added support for constants/functions etc. for batching system.
 ```A new issue should be opened since the Restart capability does not work with batching```

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Optimizers/GeneticAlgorithm.py [hunk 1]
```diff
@@ -432,12 +441,16 @@ def _useRealization(self, info, rlz):
     #        This can be simplified in the near future in GradientDescent, SimulatedAnnealing, and here in GA
     for index,individual in enumerate(offSprings):
       newOpt = individual
-      opt = objectiveVal[index]
+      opt = {self._objectiveVar:objectiveVal[index]}
+      for p,v in constraintData.items():
+        opt[p] = v[index]
+      #opt = objectiveVal[index]
```
Comment (wangcj05) [inline framework/Optimizers/GeneticAlgorithm.py:447]: remove the commented line?

File: tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMinwRep.xml [hunk 2]
```diff
@@ -92,9 +92,9 @@
       <variable name="x3">
         <distribution>uniform_dist_Repl_1</distribution>
       </variable>
-
       <objective>ans</objective>
       <TargetEvaluation class="DataObjects" type="PointSet">optOut</TargetEvaluation>
+      <constant name="unused">1.23456789</constant>
```
Comment (wangcj05) [inline tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMinwRep.xml:97]: please update the TestInfo regarding the changes.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
some minor comments for your consideration @aalfonsi

File: framework/Optimizers/GeneticAlgorithm.py (hunk 1 above)
Comment (aalfonsi) [reply framework/Optimizers/GeneticAlgorithm.py]: removed

File: tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMinwRep.xml (hunk 2 above)
Comment (aalfonsi) [reply tests/framework/Optimizers/GeneticAlgorithms/discrete/unconstrained/testGAMinwRep.xml:100]: added

**Comment by aalfonsi** [comment]
@wangcj05  Addressed your comments

**Review by wangcj05 — APPROVED** [review]
changes are good to me.

**Comment by wangcj05** [comment]
@Jimmy-INL @mandd do you have any comments?

**Comment by wangcj05** [comment]
This PR is good and checklist is satisfied. An email is needed to send to the users. @aalfonsi

**Comment by aalfonsi** [comment]
@wangcj05 created issue at #1668

**Pull request merged**